pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/TimeRangeSegmentPruner.java

```diff
@@ -0,0 +1,348 @@
+
+    List<Long> startTimes = new ArrayList<>();
+    List<Long> endTimes = new ArrayList<>();
+
+    for (int i = 0; i < segments.size(); i++) {
+      String segment = segments.get(i);
+
+      long[] range = extractStartEndTimeMSFromSegmentZKMetaZNRecord(segment, znRecords.get(i));
+      startTimes.add(range[0]);
+      endTimes.add(range[1]);
+    }
+
+    _timeRangeMSToSegmentSearchTree = new IntervalST<String>(startTimes, endTimes, segments);
+  }
+
+
+  private long[] extractStartEndTimeMSFromSegmentZKMetaZNRecord(String segment, @Nullable ZNRecord znRecord) {
+    long[] range = {MIN_START_TIME, MAX_END_TIME};
+    // Segments without metadata or with invalid time range will be set with [min_start, max_end] and will not be pruned
+    if (znRecord == null) {
+      LOGGER.warn("Failed to find segment ZK metadata for segment: {}, table: {}", segment, _tableNameWithType);
+      return range;
+    }
+
+    long startTime = znRecord.getLongField(CommonConstants.Segment.START_TIME, -1);
+    long endTime = znRecord.getLongField(CommonConstants.Segment.END_TIME, -1);
+    if (startTime < 0 || endTime < 0 || startTime > endTime) {
+      LOGGER.warn("Failed to find valid end time for segment: {}, table: {}", segment, _tableNameWithType);
+      return range;
+    }
+
+    TimeUnit timeUnit = znRecord.getEnumField(CommonConstants.Segment.TIME_UNIT, TimeUnit.class, TimeUnit.DAYS);
+    range[0] = timeUnit.toMillis(startTime);
+    range[1] = timeUnit.toMillis(endTime);
+    return range;
+  }
+
+  @Override
+  public void refreshSegment(String segment) {
+    Set<String> segments;
+    if (_timeRangeMSToSegmentSearchTree == null) {
+      segments = new HashSet<>();
+    } else {
+      segments = _timeRangeMSToSegmentSearchTree.getAllValues();
+    }
+    if (!segment.contains(segment)) {
+      segments.add(segment);
+    }
+
+    updateTimeRangeMSToSegmentSearchTree(segments);
+  }
+
+  /**
+   * NOTE: Pruning is done by searching _timeRangeToSegmentSearchTree based on request time range and check if the results
+   *       are in the input segments. By doing so we will have run time O(M * logN) (N is the # of all online segments,
+   *       M is the # of qualified intersected segments).
+   */
+  @Override
+  public Set<String> prune(BrokerRequest brokerRequest, Set<String> segments) {
+    // The pruned order may be different from the input order
+    if (_timeRangeMSToSegmentSearchTree == null) {
+      return segments;
+    }
+    FilterQueryTree filterQueryTree = RequestUtils.generateFilterQueryTree(brokerRequest);
+    if (filterQueryTree == null) {
+      return segments;
+    }
+    List<Interval> ranges = getFilterTimeRange(filterQueryTree);
+    if (ranges == null) { // cannot prune based on time for input request
+      return segments;
+    }
+    if (ranges.size() == 0) { // invalid query time range
+      return Collections.emptySet();
+    }
+
+    Set<String> prunedSegments = new HashSet<>();
```

Comment:
Actually the reverse is true right? The segment list retrieved from BST is the one that will be used. Consider changing the name to selectedSegments or something else

Reply:
I'll rename this.